src/libsyntax/feature_gate/check.rs

```diff
@@ -331,19 +327,6 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
         if let Some((.., AttributeGate::Gated(_, name, descr, has_feature))) = attr_info {
             gate_feature_fn!(self, has_feature, attr.span, name, descr, GateStrength::Hard);
         }
-        // Check input tokens for built-in and key-value attributes.
-        match attr_info {
-            // `rustc_dummy` doesn't have any restrictions specific to built-in attributes.
-            Some((name, _, template, _)) if name != sym::rustc_dummy =>
-                check_builtin_attribute(self.parse_sess, attr, name, template),
-            _ => if let Some(TokenTree::Token(token)) =
-                    attr.get_normal_item().tokens.trees().next() {
-                if token == token::Eq {
-                    // All key-value attributes are restricted to meta-item syntax.
-                    attr.parse_meta(self.parse_sess).map_err(|mut err| err.emit()).ok();
-                }
-            }
-        }
```

Comment:
So, I think I mentioned this somewhere before - this code should stay here if we want to keep the current behavior (and I'd prefer not to change observable behavior in this PR).

The function `pub fn check_attribute` above calls this `visit_attribute`, and is then called from `syntax_expand/expand.rs`.
Yes, the post-expansion visitor has to be called during expansion because some fragments of expanded AST (inert attributes on macro invocations, cc #63221) disappear later while not being unconfigured.

If there's a cyclic dependency, then some trick similar to `cfg` in modules needs to be employed (except that passing a single trait object may be better than *two* fn pointers).
Or we can continue keeping this code in libsyntax if that's possible.